source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
+			case keep_ops.KEEP_LOWEST:
+				trueDieLength = this.#selectNum;
+				break;
+			case keep_ops.DROP:
+			case keep_ops.DROP_LOWEST:
+			case keep_ops.DROP_HIGHEST:
+				trueDieLength -= this.#selectNum;
+				break;
+		}
+		return trueDieLength;
+	}
+
+	toString(frac = false) { // TODO
+		if (this.#dieList.length == 0) {
+			return frac ? `0/0` : `0`;
+		}
+		var isKeep = this.#selectOp.startsWith('k'),
+		dieMap = this.#dieList.map(result => { return { roll: result, keep: !isKeep } });
+		var skipIndices = [], repeats = this.#selectNum;
+		switch (this.#selectOp) {
+			case keep_ops.KEEP:
+			case keep_ops.KEEP_HIGHEST:
+				while (repeats > 0) {
+					var highIndex = -1;
+					var highValue = -1;
+					for (var i = 0; i < dieMap.length; i++) {
+						if (!skipIndices.includes(i) && dieMap[i].roll.getResult() > highValue) {
+							highValue = dieMap[i].roll.getResult();
+							highIndex = i;
+						}
+					}
+					dieMap[highIndex].keep = true;
+					skipIndices.push(highIndex);
+					repeats--;
+				}
+				break;
+			case keep_ops.KEEP_LOWEST:
+				while (repeats > 0) {
+					var lowIndex = -1;
+					var lowValue = Number.POSITIVE_INFINITY;
+					for (var i = 0; i < dieMap.length; i++) {
+						if (!skipIndices.includes(i) && dieMap[i].roll.getResult() < lowValue) {
+							lowValue = dieMap[i].roll.getResult();
+							lowIndex = i;
+						}
+					}
+					dieMap[lowIndex].keep = true;
+					skipIndices.push(lowIndex);
+					repeats--;
+				}
+				break;
+			case keep_ops.DROP:
+			case keep_ops.DROP_LOWEST:
+				while (repeats > 0) {
+					var lowIndex = -1;
+					var lowValue = Number.POSITIVE_INFINITY;
+					for (var i = 0; i < dieMap.length; i++) {
+						if (!skipIndices.includes(i) && dieMap[i].roll.getResult() < lowValue) {
+							lowValue = dieMap[i].roll.getResult();
+							lowIndex = i;
+						}
+					}
+					dieMap[lowIndex].keep = false;
+					skipIndices.push(lowIndex);
+					repeats--;
+				}
+				break;
+			case keep_ops.DROP_HIGHEST:
+				while (repeats > 0) {
+					var highIndex = -1;
+					var highValue = -1;
+					for (var i = 0; i < dieMap.length; i++) {
+						if (!skipIndices.includes(i) && dieMap[i].roll.getResult() > highValue) {
+							highValue = dieMap[i].roll.getResult();
+							highIndex = i;
+						}
+					}
+					dieMap[highIndex].keep = false;
+					skipIndices.push(highIndex);
+					repeats--;
+				}
+				break;
+		}
+		return `(${dieMap.map(dieInfo => dieInfo.keep ? dieInfo.roll.toString(frac) : `~~${dieInfo.roll.toString(frac)} dropped~~${frac ? '0/0' : '0'}`).join(' + ')})`;
+	}
+
+	toResultArray() {
+		return this.#dieList.flatMap(r => r.getResult());
+	}
+}
+
+class ResultBundle {
+	#resultset
+	#extraText
+
+	constructor(rs, et = ""){
+		var resultSet;
+		if (rs instanceof ResultSet) { // If it is a result set, just use it as-is
+			resultSet = rs;
+		} else { // Parse the string into a roll set
+			resultSet = ResultSet.parse(rs);
+		}
+		this.#resultset = resultSet;

```

Comment:
```suggestion
		this.#resultset = rs instanceof ResultSet ? rs : ResultSet.parse(rs);
```
Perhaps paradoxically, I think a ternary is good here due to it removing need for an intermediate variable.

Reply:
```suggestion
		if (rs instanceof ResultSet) { // If it is a result set, just use it as-is
			this.#resultset = rs;
		} else { // Parse the string into a roll set
			this.#resultset = ResultSet.parse(rs);
		}
```
A version in the name of avoiding ternaries.

Reply:
That ternary is sweet, and "convert to right type if not already" is a pretty clean ternary use case. I'll definitely do this change.